A29-xds-tls-security.md

```diff
@@ -0,0 +1,340 @@
+
+#### gRPC Server (Listener) Side Changes
+
+Adding xDS-based transport security on the server side requires server-side xDS support
+first, which currently does not exist. Adding xDS support to gRPC servers
+is a separate design activity and is out of scope for this proposal. However for further
+discussion, we assume the support is in place.
+
+Once a gRPC server is created with a `XdsServerCredentials` and the appropriate
+xDS-related server options, the server is set up to receive the relevant LDS responses
+from the xDS control plane containing the
+[Listener](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/config/listener/v3/listener.proto#L31)
+resource. When transport security (i.e. TLS/mTLS) is enabled, the resource will include
+[DownstreamTlsContext](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L58) containing the TLS configuration needed for all the incoming connections
+to that listener. The local identity certificate
+information is in [tls_certificate_certificate_provider](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L169). For mTLS, it will also include
+root-certificate information in the validation context fields - either
+[validation_context_certificate_provider](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L190) or [combined_validation_context.validation_context_certificate_provider](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L145).
+
+Once the gRPC Server (listener) receives an LDS update that contains the DownstreamTlsContext,
+it will ask the [Certificate Provider Store](#certificate-provider-store) to find or create a
+[Certificate Provider](#certificate-provider) for each of the two `certificate_provider`
+values received. If these two values turn out to be the same, the same Certificate
+Provider will be used in both the cases. The [Distributor](#distributor)s
+associated with the Certificate Providers are used to receive Certificate updates and
+provide them to each incoming connection's security connector.
+
+![gRPC xDS Security Server Flow Diagram](A29_graphics/grpc_xds_security_server_flow.png)
+
+## Supporting SDS (for interop with Envoy)
+
+Envoy currently uses SDS to acquire certificates from an SDS server (which eventually
+gets them from a Certification Authority). Current Envoy deployments receive
+SDS configuration from the xDS control plane. For gRPC to seamlessly interoperate with
+Envoy, it needs to be able to consume the same SDS config. It is possible to support
+SDS within this plugin architecture where an SDS client can be implemented as a
```

Comment:
Let's not be proscriptive here.  We haven't yet agreed on whether we would implement this as a CertificateProvider plugin or as a separate code path.  And in the context of this document, it doesn't really matter which of those routes we choose.  I think the important thing to say here is that we do intend to eventually add SDS support as well.